rust/src/encodings/plain.rs

```diff
@@ -238,13 +220,41 @@ impl<'a> Decoder for PlainDecoder<'a> {
         if indices.is_empty() {
             return Ok(new_empty_array(self.data_type));
         }
+        let block_size = self.reader.prefetch_size() as u32;
+        let byte_width = self.data_type.byte_width() as u32;
+
+        let mut chunk_ranges = vec![];
+        let mut start: u32 = 0;
+
+        for idx in 0..(indices.len() - 1) as u32 {
+            if indices.value(idx as usize + 1) * byte_width
+                > indices.value(start as usize) * byte_width + block_size
+            {
+                chunk_ranges.push(start..idx + 1);
+                start = idx + 1;
+            }
+        }
+        // Remaining
+        if start < indices.len() as u32 {
+            chunk_ranges.push(start..indices.len() as u32);
+        }
 
-        // TODO: optimize read for sparse indices later.
-        let start = indices.value(0);
-        let end = indices.value(indices.len() - 1);
-        let array = self.get(start as usize..end as usize + 1).await?;
-        let adjusted_offsets = subtract_scalar(indices, start)?;
-        Ok(take(&array, &adjusted_offsets, None)?)
+        let arrays = stream::iter(chunk_ranges)
+            .map(|cr| async move {
+                let index_chunk = indices.slice(cr.start as usize, cr.len());
+                let request: &UInt32Array = as_primitive_array(&index_chunk);
+
+                let start = request.value(0);
+                let end = request.value(request.len() - 1);
+                let array = self.get(start as usize..end as usize + 1).await?;
```

Comment:
not sure what's Rust best practice - is it better to do ..= or .. and +1?

Reply:
`..=` is actually `inclusive`, where  `a..=b` equals to `a..b-1`?